Assembler.Core/IHandler.cs

```diff
@@ -1,7 +1,14 @@
-﻿namespace Assembler.Core
+﻿using System;
+using Assembler.Core.Entities;
+
+namespace Assembler.Core
 {
-    public interface IHandler
+    public interface IFrameHandler<in TFrame, out TMessage>
+        where TFrame : BaseFrame
+        where TMessage : BaseMessageInAssembly
     {
-        void Handle(IMessage message);
+        void Handle(TFrame message);
+
+        event Action<TMessage> MessageAssemblyFinished;
```

Comment:
I'm thinking about releasing the message (whether it is expired or finished) - I dont think the Assembler should subscribe to the handlers directly. It should subscribe to some IMessageReleaser, which looks like that:

![image](https://user-images.githubusercontent.com/62211836/77541220-045fb480-6ead-11ea-8830-aadd16dabd4a.png)

Release by frame flow -> BaseFrameHandler calls Release
Release by timeout flow -> cache calls Release

then the releaser gets and removes the messageInAssembly from the cache, set the ReleaseReason, and raise the event.

The Assembler subscribe it's `ReleaseMessage` to the event (and we can delete the `ReleaseExpiredMessage` method).

What do you think?

Reply:
I thought  about it and if we are using ITimedBaseCache it might be problematic, because the implementation will remove the item from cache by itself...

But even though, I think that there should be single place that get the Id of the message (or the message itself) and the reason, set the reason in the message and raise it's event. Again, subscribing the Assembler directly to the handlers isnt right, it should not know them.

Reply:
I want to talk to you about it